tokio/src/fs/read_uring.rs

```diff
@@ -0,0 +1,137 @@
+            // The buffer might be an exact fit. Let's read into a probe buffer
+            // and see if it returns `Ok(0)`. If so, we've avoided an
+            // unnecessary increasing of the capacity. But if not, append the
+            // probe buffer to the primary buffer and let its capacity grow.
+            let (size_read, r_fd, r_buf) = small_probe_read(fd, buf, &mut offset).await?;
+
+            if size_read == 0 {
+                return Ok(r_buf);
+            }
+
+            buf = r_buf;
+            fd = r_fd;
+        }
+
+        // buf is full, need more capacity
+        if buf.len() == buf.capacity() {
+            buf.try_reserve(PROBE_SIZE)?;
+        }
+
+        // prepare the spare capacity to be read into
+        let buf_len = usize::min(buf.spare_capacity_mut().len(), MAX_READ_SIZE);
+
+        // buf_len cannot be greater than u32::MAX because MAX_READ_SIZE
+        // is less than u32::MAX
+        let read_len = u32::try_from(buf_len).expect("buf_len must always fit in u32");
+
+        // read into spare capacity
+        let (size_read, r_fd, r_buf) = op_read(fd, buf, &mut offset, read_len).await?;
+
+        if size_read == 0 {
+            return Ok(r_buf);
+        }
+
+        fd = r_fd;
+        buf = r_buf;
+    }
+}
+
+async fn small_probe_read(
+    fd: OwnedFd,
+    mut buf: Vec<u8>,
+    offset: &mut u64,
+) -> io::Result<(u32, OwnedFd, Vec<u8>)> {
+    let read_len = PROBE_SIZE_U32;
+
+    let mut temp_arr = [0; PROBE_SIZE];
+    // we don't call this function if buffer's length < PROBE_SIZE
+    let back_bytes_len = buf.len() - PROBE_SIZE;
+
+    temp_arr.copy_from_slice(&buf[back_bytes_len..]);
+
+    // We're decreasing the length of the buffer and len is greater
+    // than PROBE_SIZE. So we can read into the discarded length
+    buf.truncate(back_bytes_len);
+
+    let (size_read, r_fd, mut r_buf) = op_read(fd, buf, offset, read_len).await?;
+    // If `size_read` returns zero due to reasons such as buffer's exact fit,
+    // then this `try_reserve` does not perform allocation.
+    r_buf.try_reserve(PROBE_SIZE)?;
+    r_buf.splice(back_bytes_len..back_bytes_len, temp_arr);
+
+    Ok((size_read, r_fd, r_buf))
+}
+
+async fn op_read(
+    mut fd: OwnedFd,
+    mut buf: Vec<u8>,
+    offset: &mut u64,
+    mut read_len: u32,
+) -> io::Result<(u32, OwnedFd, Vec<u8>)> {
+    loop {
+        let (res, r_fd, r_buf) = Op::read(fd, buf, read_len, *offset).await;
+
+        match res {
+            Err(e) if e.kind() == ErrorKind::Interrupted => {
+                buf = r_buf;
+                fd = r_fd;
+            }
+            Err(e) => return Err(e),
+            Ok(size_read) => {
+                *offset += size_read as u64;
+                read_len -= size_read;
+
+                if read_len == 0 || size_read == 0 {
+                    return Ok((size_read, r_fd, r_buf));
+                }
+
+                buf = r_buf;
+                fd = r_fd;
+            }
```

Comment:
I do not think this makes sense.
```suggestion
            Ok(size_read) => return Ok((size_read, r_fd, r_buf)),
```
The current logic means that unless the buffer fits the file exactly, we perform the final read of length zero twice:

1. We call `op_read`, which performs a partial read and then a length-zero EOF read. It returns a non-zero length to the caller.
2. Since the caller got a non-zero length, it now needs to call `op_read` again to check for EOF.

Checking for EOF twice does not make sense.